ykllvmwrap/src/jitmodbuilder.cc

```diff
@@ -288,15 +290,53 @@ void dumpValueAndExit(const char *Msg, Value *V) {
   exit(EXIT_FAILURE);
 }
 
-// A function name and basic block index pair that identifies a block in the
-// AOT LLVM IR.
+// A function name and basic block index pair that identifies a mappable block
+// in the AOT LLVM IR.
 struct IRBlock {
   // A non-null pointer to the function name.
   char *FuncName;
   // The index of the block in the parent LLVM function.
   size_t BBIdx;
 };
 
+// An unmappable region of code spanning one or more machine blocks.
+struct UnmappableRegion {
+  // The effect executing the unmappable block had on the stack depth. This is
+  // number of frames (not the sie of the stack in bytes).
+  //
+  // Since each UnmappableRegion in a trace is followed by a mappable IRBlock,
+  // this value can be used to determine if the unmappable code:
+  //
+  //  - returned back to mappable code (StackAdjust < 0).
+  //  - called deeper into mappable code (StackAdjust > 0).
+  //  - "fell through" into mappable code (StackAdjust ==  0).
+  ssize_t StackAdjust;
+};
+
+class TraceLoc {
+  std::variant<IRBlock, UnmappableRegion> Loc;
+
+public:
+  TraceLoc(std::variant<IRBlock, UnmappableRegion> Loc) : Loc(Loc) {}
+
+  UnmappableRegion *getUnmappableRegion() {
+    return std::get_if<UnmappableRegion>(&Loc);
+  }
+
+  IRBlock *getMappedBlock() { return std::get_if<IRBlock>(&Loc); }
+
+  void dump() {
+    if (IRBlock *IRB = std::get_if<IRBlock>(&Loc)) {
+      errs() << "IRBlock(Func=" << IRB->FuncName << ", BBIdx=" << IRB->BBIdx
+             << ")\n";
+    } else if (UnmappableRegion *U = std::get_if<UnmappableRegion>(&Loc)) {
```

Comment:
This me clearer as `if (...) {...} else { assert(...); ...}` without the `abort`? Unless you're worried that this *might* happen even in release mode?

Reply:
Good idea.